benches/latency.rs

```diff
@@ -0,0 +1,218 @@
+                Ok(futures::Async::Ready(Some(t))) => { self.buf = Some(t) },
+                Ok(futures::Async::NotReady) => return Ok(futures::Async::NotReady),
+                Err(e) => return try_nb!(Err(e)),
+            }
+        }
+    }
+}
+#[bench]
+fn udp_echo_latency(b: &mut Bencher) {
+    let server_addr= "127.0.0.1:7398".to_string();
+    let server_addr = server_addr.parse::<SocketAddr>().unwrap();
+    let client_addr= "127.0.0.1:7399".to_string();
+    let client_addr = client_addr.parse::<SocketAddr>().unwrap();
+
+    let stop = Arc::new(AtomicBool::new(false));
+    let stop2 = stop.clone();
+
+    let child = thread::spawn(move || {
+        let mut l = Core::new().unwrap();
+        let handle = l.handle();
+
+        let socket = tokio_core::net::UdpSocket::bind(&server_addr, &handle).unwrap();
+
+        let server = EchoServer::new(socket, stop);
+
+        l.run(server).unwrap();
+    });
+
+    // TODO: More reliable way to bind server socket and start server
+    // first
+    thread::sleep(Duration::from_millis(100));
+
+    let client = std::net::UdpSocket::bind(client_addr).unwrap();
+
+    let mut buf = [0u8; 1000];
+
+    // warmup phase; for some reason initial couple of
+    // rounds is much slower
+    //
+    // TODO: Describe the exact reasons; caching? branch predictor? lazy closures?
+    for _ in 0..1000 {
+        client.send_to(&buf, &server_addr).unwrap();
+        let _ = client.recv_from(&mut buf).unwrap();
+    }
+
+    b.iter(|| {
+        client.send_to(&buf, &server_addr).unwrap();
+        let _ = client.recv_from(&mut buf).unwrap();
+    });
+
+    // Stop the server; TODO: Use better method
+    stop2.store(true, Ordering::SeqCst);
+    thread::sleep(Duration::from_millis(1));
+    client.send_to(&buf, &server_addr).unwrap();
+
+    child.join().unwrap();
+}
+
+#[bench]
+fn channel_latency(b: &mut Bencher) {
+    let stop = Arc::new(AtomicBool::new(false));
+    let stop2 = stop.clone();
+
+    // TODO: Any way to start Loop on a separate thread and yet get
+    // a tokio_core::channel to it?
+    let (tx, rx) = std::sync::mpsc::channel();
+
+    let child = thread::spawn(move || {
+        let mut l = Core::new().unwrap();
+        let handle = l.handle();
+
+        let (in_tx, in_rx) = tokio_core::channel::channel(&handle).unwrap();
+        let (out_tx, out_rx) = tokio_core::channel::channel(&handle).unwrap();
+
+        let server = ChanServer::new(out_tx, in_rx, stop);
+
+        tx.send((in_tx, out_rx)).unwrap();
+        l.run(server).unwrap();
+    });
+
+    let (in_tx, out_rx) = rx.recv().unwrap();
+
+    // TODO: More reliable way to bind server socket and start server
+    // first
+    thread::sleep(Duration::from_millis(100));
+
+    let mut rx_iter = out_rx.wait();
+
+    // warmup phase; for some reason initial couple of
+    // rounds is much slower
+    //
+    // TODO: Describe the exact reasons; caching? branch predictor? lazy closures?
```

Comment:
Yes. I've actually added it to make sure server had time to start up and can respond immediately. I'll double check and investigate more.